File: c2/src/domain/entities.py
from enum import IntEnum, auto
from uuid import uuid4
from pydantic import BaseModel, Field
from strenum import PascalCaseStrEnum


class AccessLevel(IntEnum):
    Nil = 0
    UserRead = 1
    UserWrite = 2
    UserExecute = 3
    RooWriteRead = 1
    RooWriteWrite = 2
    RooWriteExecute = 3


class Entity(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    name: str
    aliases: list[str] = []


class C2(Entity):
    pass


class Asset(BaseModel):
    pass


class System(Entity):
    # TODO maybe add UID, GID, arch,
    name: str
    ip: str | None = None
    os: str = "linux"
    username: str | None = None
    version: str | None = None
    transport: str | None = None
    ports: dict[int | str, int] = {}  # key is either port name or port number
    access_level: AccessLevel = AccessLevel.Nil


# JWT RFC: https://datatracker.ietf.org/doc/html/rfc7519#section-4
class JWTToken(Asset):
    subject: str | None = None
    audience: list[str] = []
    issuer: str | None = None
    expires_at: int | None = None
    issued_at: int | None = None
    not_valid_before: int | None = None
    raw: str | None = None


class ServiceAccountToken(JWTToken):
    # TODO verify if issuer is indicater of K8s api server?
    namespace: str | None = None
    pod_name: str | None = None
    pod_uid: str | None = None
    serviceaccount_name: str | None = None
    serviceaccount_uid: str | None = None
    warn_after: int | None = None


# TODO differentiate between k8s resources and a IAM entity?
class ServiceAccount(Entity):
    kind: str = "ServiceAccount"
    ns: str
    name: str
    token: str | ServiceAccountToken | None = Field(None, exclude=True)
    expires_at: int | None = None
    can: list[str] = []


class Namespace(Entity):
    kind: str = "Namespace"


class Pod(System):
    kind: str = "Pod"
    name: str
    ns: str | Namespace | None = None
    service_account: ServiceAccount | None = None 


class Service(Entity):
    kind: str = "Service"
    name: str
    ip: str | None = None
    ports: dict[int | str, int] = {}  # key is either port name or port number
    ns: str | Namespace | None = None


class ApiServer(System):
    kind: str = "KubeApiServer"
    name: str = "API Server"
    ns: str = "kube-system"


class Relation(BaseModel):
    name: str
    source: str
    destination: str
    data: str | None = None


class SystemState(PascalCaseStrEnum):
    Unknown = auto()
    Known = auto()
    Reachable = auto()
    Compromised = auto()


class Topology(BaseModel):
    entities: list[Entity] = []
    relations: list[Relation] = []

File: Ran/src/domain/entities.jl
abstract type Entity end

abstract type AbstractSystem <: Entity end

SystemId = AbstractString

Base.@kwdef struct System <: AbstractSystem
    id:: SystemId = string(uuid4())
    name::AbstractString
    os::Union{AbstractString,Nothing} = nothing
    accessLevel :: Union{AccessLevel,Nothing} = nothing
end

Base.@kwdef struct Listener <: AbstractSystem
    id:: AbstractString = string(uuid4())
    kind :: String = "Listener"
    name::AbstractString
    port :: Int
    protocol :: Union{AbstractString, Nothing} = nothing
    # os::Union{AbstractString,Nothing} = nothing
    # accessLevel :: Union{AccessLevel,Nothing} = nothing
end

Base.@kwdef struct Namespace <: Entity
    id :: AbstractString = string(uuid4())
    kind:: String = "Namespace"
    name:: AbstractString
end

abstract type AbstractToken <: Asset end


# JWT RFC: https://datatracker.ietf.org/doc/html/rfc7519#section-4
Base.@kwdef struct JWTToken <: AbstractToken
    subject:: Union{AbstractString, Nothing} = nothing
    audience:: Vector{AbstractString} = []
    issuer:: Union{AbstractString, Nothing} = nothing
    expiresAt:: Union{Int, Nothing} = nothing
    issuedAt:: Union{Int, Nothing} = nothing
    notValidBefore:: Union{Int, Nothing} = nothing
    raw:: Union{AbstractString, Nothing} = nothing
end


Base.@kwdef struct ServiceAccountToken <: AbstractToken
    jwtToken:: JWTToken # ServiceAccountToken is just a JWT token 
    # TODO verify if issuer is indicater of K8s api server?
    namespace:: Union{AbstractString, Nothing} = nothing
    podName:: Union{AbstractString, Nothing} = nothing
    podUid:: Union{AbstractString, Nothing} = nothing
    serviceAccountName:: Union{AbstractString, Nothing} = nothing
    serviceAccountUid:: Union{AbstractString, Nothing} = nothing
    warnAfter:: Union{Int, Nothing} = nothing
end


Base.@kwdef struct ServiceAccount <: Entity
    id :: AbstractString = string(uuid4())
    kind:: String = "ServiceAccount"
    ns:: Union{AbstractString, Namespace, Nothing} = nothing
    name:: AbstractString
    token:: Union{AbstractString, ServiceAccountToken, Nothing} = nothing
    expiresAt:: Union{Int, Nothing} = nothing
    can:: Vector{AbstractString} = []
end

Base.@kwdef struct Pod <: AbstractSystem
    id :: AbstractString = string(uuid4())
    kind:: String = "Pod"
    name:: AbstractString
    ns:: Union{AbstractString, Namespace, Nothing} = nothing
    serviceAccount:: Union{ServiceAccount, Nothing} = nothing
end


Base.@kwdef struct ApiServer <: AbstractSystem
    id :: AbstractString = string(uuid4())
    kind:: String = "KubeApiServer"
    name:: String = "API Server"
    ns:: Union{AbstractString, Namespace} = "kube-system"
    ip:: Union{AbstractString, Nothing} = nothing
    ports:: Dict{AbstractString, Union{Int, AbstractString}} = Dict()
end

Base.@kwdef struct Service <: Entity
    id :: AbstractString = string(uuid4())
    kind:: String = "Service"
    name:: AbstractString
    ns:: Union{AbstractString, Namespace, Nothing} = nothing
    ip:: Union{AbstractString, Nothing} = nothing
    ports:: Dict{AbstractString, Union{Int, AbstractString}} = Dict()
end